### PR TITLE
Refactor common

### LINKED ISSUE #503: [Proposal] RL Common / Baselines Common Package
With @hill-a , we came up to the same conclusion about things in the common folder: a lot is not specific to this library and could be separated in a stand-alone python package.

Doing so would:
- allow people to develop/build things on that package without depending on a specific backend (i.e. this removes tensorflow dependency)
- separate what is related to rl in general vs stable-baselines
- ease the transition between tf 1 and tf 2 (#366) as we could decide to have two separate versions of SB instead of maintaining a version with tf1 and tf2 support at the same time (which seems complicated especially because of the removal of tf.contrib)

As a personal example, I have now an internal (minimal) pytorch version of stable-baselines (I plan to release it at some point but it is used for my own research purposes for now) where I had to copy a lot of what was present in the common folder.

I would like to have your thoughts on that ;)
@erniejunior @AdamGleave @Miffyli 

Suggestions for the name of the package are also welcomed, for now we have:
- baselines-common (by @hill-a)
- rl-common (my proposition)

the repo would be under the stable-baselines team organization: https://github.com/Stable-Baselines-Team

### PR DESCRIPTION
**Pull request by Miffyli**
Refactoring `common` to include most shared code between learning algorithms. 

## Description
Move (almost) any shared code out from algorithm-specific directories into commons, e.g:
* `a2c/utils.py` defined most of the TF layers and utilities. Now in `common/tf_layers.py`
* `ReplayBuffer` was defined under DQN but also used by DDPG. Moved to `common/replay_buffer.py` .
* Some math utils like `safe_mean` moved from under PPO2 to `common/math_util.py`

Also removed unused code `common/tf_util.py, calc_entropy_softmax`.

PPO1/TRPO still share code between themselves (defined in PPO1). This should also be moved in commons, but at the same time they are the only ones using this very specific type of code (trajectory generation).

Things to-do and discuss:
* Unifying some of the code. E.g. there are now two different types of parameter schedulers in `common/schedulers.py`. 

## Motivation and Context
closes #503 

While there is a `common` directory for shared code between learning algorithms, bunch of shared code existed under learning algorithms that was imported elsewhere code (e.g. Half of the TF definitions were under `a2c/utils.py`). As discussed in #503, it would make more sense to have learning algorithms import shared code from `common`, and `common` should not depend on per-algorithm definitions.

This also aims to keep all backend (Tensorflow) related code few files, now contained in `tf_util.py` and `tf_layers.py` under `common`. This should help a tiny bit when transitioning to other backends. 

## Types of changes
- [x] Refactoring (no changes to documentation)

## Checklist:
- [x] I've read the [CONTRIBUTION](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the [changelog](https://github.com/hill-a/stable-baselines/blob/master/docs/misc/changelog.rst) accordingly (**required**).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by araffin** [comment]
Good that you started the refactoring =)
I will try to give you some feedback this week and open at the same time a PR for tf2 to have more visibility on the progress

**Comment by Miffyli** [comment]
Given that this is mostly related to (and motivated by) the transition to new a new backend, I think we should merge this with the TF2 branch (#542) whenever it is ready for merging.

**Comment by araffin** [comment]
I would merge it with both ;) (I should have the time today to look at this one)

File: stable_baselines/a2c/utils.py [hunk 1]
```diff
@@ -1,297 +1,3 @@
...
-
-    :param input_tensor: (TensorFlow Tensor) The convolutional input tensor
-    :return: (TensorFlow Tensor) The fully connected output tensor
-    """
-    n_hidden = np.prod([v.value for v in input_tensor.get_shape()[1:]])
-    input_tensor = tf.reshape(input_tensor, [-1, n_hidden])
-    return input_tensor
-
-
 def discount_with_dones(rewards, dones, gamma):
```
Comment (araffin) [inline stable_baselines/a2c/utils.py]: Is there a reason you did not move that one?

File: stable_baselines/common/schedules.py [hunk 2]
```diff
@@ -106,3 +106,149 @@ def __init__(self, schedule_timesteps, final_p, initial_p=1.0):
...
+    eps1 = 0.75
+    eps2 = 0.25
+    if 1 - progress < eps1:
+        if 1 - progress < eps2:
+            return eps2 * 0.5
+        return eps1 * 0.1
+    return 1 - progress
+
+
+SCHEDULES = {
```
Comment (araffin) [inline stable_baselines/common/schedules.py]: I think we should also take advantage of the migration to unify the scheduling, I would go for something like in PPO2/SAC/TD3, where you can pass a callable.

File: stable_baselines/common/tf_util.py [hunk 3]
```diff
@@ -335,3 +466,38 @@ def _getter(getter, name, *args, **kwargs):
         val = getter(name, *args, **kwargs)
         return val
     return _getter
+
+
+# ================================================================
+# Logging
+# ================================================================
+
+def total_episode_reward_logger(rew_acc, rewards, masks, writer, steps):
```
Comment (araffin) [inline stable_baselines/common/tf_util.py]: Note for myself: we should fix this one, it is not working properly when n_envs > 0 for a while now...

File: stable_baselines/a2c/utils.py (hunk 1 above)
Comment (Miffyli) [reply stable_baselines/a2c/utils.py]: This one was only imported by A2C so in that sense it was in its right place (and this is probably third-ish implementation of that whole function). It could be moved inside main a2c code though to get rid of this util file.

**Comment by Miffyli** [comment]
Regarding this PR: I am starting to have conflicted feelings about moving things to `common`. One one hand it makes intuitive sense to have all shared code between and algorithms should not depend on each other, but on the other hand some of the current implementations are very algorithm specific (e.g. only used by PPO1 and TRPO). Doing this PR would end up with code in common which in reality is tightly tied to those specific algorithms. 

Should we continue loyal to these guidelines (algorithms import common, common does not import algorithms), or should we figure out better refactoring while doing this? Some refactoring was already discussed on Schedulers.

File: stable_baselines/deepq/__init__.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 from stable_baselines.deepq.policies import MlpPolicy, CnnPolicy, LnMlpPolicy, LnCnnPolicy
 from stable_baselines.deepq.build_graph import build_act, build_train  # noqa
 from stable_baselines.deepq.dqn import DQN
-from stable_baselines.deepq.replay_buffer import ReplayBuffer, PrioritizedReplayBuffer  # noqa
+from stable_baselines.common.replay_buffer import ReplayBuffer, PrioritizedReplayBuffer  # noqa
```
Comment (araffin) [inline stable_baselines/deepq/__init__.py]: In fact, I would call that file buffers because I would like to add a `RolloutBuffer` later (at least for v3)

File: stable_baselines/common/schedules.py (hunk 2 above)
Comment (Miffyli) [reply stable_baselines/common/schedules.py]: Got it. I will unify these to use the newer version used by PPO2/SAC/TD3, and try to retain some of the existing schedules.

Comment (Miffyli) [reply stable_baselines/common/schedules.py]: After refreshing my head on this one, I think a mix of the classical scheduler where you get `progress` in [0, 1] and return learning-rates and such is better. The PPO2/SAC/TD3 one requires users to play around with the number of steps, which is not always clear what one "step" is, depending on the algorithm.

I suggest we skip this part for now and focus on finishing rest of the refactoring (Yesyes, me lazy).

Comment (araffin) [reply stable_baselines/common/schedules.py]: ok ;)

**Comment by Miffyli** [comment]
Alright, now everything should be done. The only remaining "oddball" is GAIL importing from TRPO, but since these two are too intimately linked I decided to leave them as it is (and these algos should go to Adam's repository, as I understood).

@araffin 
If good for merging, we can first merge e.g. #644 and I can sort out the merge conflicts and whatnot before merging this one :)

File: docs/misc/changelog.rst [hunk 5]
```diff
@@ -57,6 +57,13 @@ Others:
 - Removed redundant return value from ``a2c.utils::total_episode_reward_logger``. (@shwang)
 - Cleanup and refactoring in ``common/identity_env.py`` (@shwang)
 - Added a Makefile to simplify common development tasks (build the doc, type check, run the tests)
+- Removed unused code in `common/tf_util.py` (`calc_entropy_softmax`)
+- Refactoring:
+
+   - Algorithms no longer import from each other.
+   - `common` no longer import from algorithms.
+   - Moved shared code to new files `common/math_util.py`, `common/buffers.py` and `common/tf_layers.py`.
```
Comment (araffin) [inline docs/misc/changelog.rst]: Those are breaking changes, no?

Maybe it would be good to list which function were moved and where

Comment (Miffyli) [reply docs/misc/changelog.rst]: Right, good point. Somehow I completely missed that ^^. I will document this.

Comment (Miffyli) [reply docs/misc/changelog.rst]: Updated changelog with such a list.
